src/librustc_codegen_llvm/coverageinfo/mapgen.rs

```diff
@@ -0,0 +1,197 @@
+use crate::llvm;
+
+use crate::common::CodegenCx;
+use crate::coverageinfo;
+
+use log::debug;
+use rustc_codegen_ssa::coverageinfo::map::*;
+use rustc_codegen_ssa::traits::{BaseTypeMethods, ConstMethods, MiscMethods};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_middle::bug;
+use rustc_span::{FileName, RealFileName, SourceFile};
+
+use std::collections::BTreeMap;
+use std::ffi::CString;
+
+// FIXME(richkadel): Complete all variations of generating and exporting the coverage map to LLVM.
+// The current implementation is an initial foundation with basic capabilities (Counters, but not
+// CounterExpressions, etc.).
+
+/// Generates and exports the Coverage Map.
+///
+/// This Coverage Map complies with Coverage Mapping Format version 3 (zero-based encoded as 2),
+/// as defined at [LLVM Code Coverage Mapping Format](https://github.com/rust-lang/llvm-project/blob/llvmorg-8.0.0/llvm/docs/CoverageMappingFormat.rst#llvm-code-coverage-mapping-format)
+/// and published in Rust's current (July 2020) fork of LLVM. This version is supported by the
+/// LLVM coverage tools (`llvm-profdata` and `llvm-cov`) bundled with Rust's fork of LLVM.
+///
+/// Consequently, Rust's bundled version of Clang also generates Coverage Maps compliant with
+/// version 3. Clang's implementation of Coverage Map generation was referenced when implementing
+/// this Rust version, and though the format documentation is very explicit and detailed, some
+/// undocumented details in Clang's implementation (that may or may not be important) were also
+/// replicated for Rust's Coverage Map.
+pub fn finalize<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>) {
+    let name_ref_i64 = cx.type_i64();
+    let mapping_data_size_u32 = cx.type_i32();
+    let func_hash_u64 = cx.type_i64();
+    let function_record_ty = cx
+        .type_struct(&[name_ref_i64, mapping_data_size_u32, func_hash_u64], /*packed=*/ true);
+
+    let mut filenames = Vec::<CString>::new();
+    let mut next_file_id = 0;
+    let mut filename_to_id = FxHashMap::<String, u32>::default();
+    let mut function_records = Vec::<&'ll llvm::Value>::new();
+
+    let mut coverage_mappings_buffer = llvm::build_byte_buffer(|coverage_mappings_buffer| {
```

Comment:
The rightward drift in here is getting heavy, maybe there's an opportunity for refactoring, not sure.

Also, a comment describing what's going on overall (without needing to read the reference docs) would probably help.

Reply:
I did a lot of refactoring to break things up better. Let me know if you have any alternative suggestions. Thanks!